### PR TITLE
fix: allow plugins to teardown to remove listeners

### LINKED ISSUE #457: Undefined document in file-download-tracking
## Expected Behavior

No error if `document` is undefined.

## Current Behavior

In a next.js app with SSR enabled, and amplitude v2.1, I'm seeing the error `document is not defined` at this line 

https://github.com/amplitude/Amplitude-TypeScript/blob/8f4ea010279fb21190a2c0595d4ae8a7d9e987ce/packages/analytics-browser/src/plugins/file-download-tracking.ts#L49

This seems similar to https://github.com/amplitude/Amplitude-TypeScript/issues/444. I have not seen this with prior versions of the SDK.

## Possible Solution

Check if `document` is defined; fall back to an empty array.

## Steps to Reproduce

I believe this would have the same reproduction as https://github.com/amplitude/Amplitude-TypeScript/issues/444. Run a next.js app with SSR, and include amplitude with file download tracking enabled.

## Environment

- JS SDK Version: `"@amplitude/analytics-browser": "2.1.0"`
- Installation Method: yarn
- Browser and Version:

### PR DESCRIPTION
**Pull request by kevinpagtakhan**
### Summary

* Adds teardown method to plugins interface
* Removes event listeners on plugin teardown
* Fixes #457

### Checklist

* [X] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

**Review by justin-fiedler — COMMENTED** [review]
LGTM

File: packages/analytics-browser/src/plugins/form-interaction-tracking.ts [hunk 1]
```diff
@@ -8,7 +8,30 @@ import {
...
+    element.addEventListener(type, handler);
+    eventListeners.push({
+      element,
+      type,
+      handler,
+    });
+  };
+  const removeClickListeners = () => {
+    eventListeners.forEach(({ element, type, handler }) => {
+      element.removeEventListener(type, handler);
```
Comment (liuyang1520) [inline packages/analytics-browser/src/plugins/form-interaction-tracking.ts:30]: Maybe `element?.removeEventListener`, not sure whether if `element` is removed from the DOM, this reference will be null or not.

**Review by liuyang1520 — APPROVED** [review]

File: packages/plugin-page-view-tracking-browser/test/page-view-tracking.test.ts [hunk 2]
```diff
@@ -290,6 +290,24 @@ describe('pageViewTrackingPlugin', () => {
     });
   });
 
+  describe('teardown', () => {
```
Comment (justin-fiedler) [inline packages/plugin-page-view-tracking-browser/test/page-view-tracking.test.ts:293]: In a recent session fix for mobile, sometimes `onExitForeground` would not get called (crash, etc). Do you see any potential for a similar issue here? 

Is there any harm in calling `teardown()` in `setup()` to be safe?

File: packages/analytics-browser/src/plugins/form-interaction-tracking.ts (hunk 1 above)
Comment (kevinpagtakhan) [reply packages/analytics-browser/src/plugins/form-interaction-tracking.ts]: fixed!

File: packages/plugin-page-view-tracking-browser/test/page-view-tracking.test.ts (hunk 2 above)
Comment (justin-fiedler) [reply packages/plugin-page-view-tracking-browser/test/page-view-tracking.test.ts:293]: Discussed in Slack.

image

**Review by justin-fiedler — APPROVED** [review]
Please also add to docs 🙏

File: packages/analytics-browser/src/plugins/form-interaction-tracking.ts [hunk 3]
```diff
@@ -8,7 +8,31 @@ import {
...
+interface EventListener {
+  element: Element;
+  type: 'change' | 'submit';
+  handler: () => void;
+}
+
 export const formInteractionTracking = (): EnrichmentPlugin => {
+  let observer: MutationObserver | undefined;
+  let eventListeners: EventListener[] = [];
+  const addEventListener = (element: Element, type: 'change' | 'submit', handler: () => void) => {
```
Comment (kelsonpw) [inline packages/analytics-browser/src/plugins/form-interaction-tracking.ts:20]: nit: I find it confusing to see a function named `addEventListener`, my brain immediately jumps to "is this calling `window.addEventListener` implicitly? Same with remove. Nbd

**Pull request merged**